cilium/CFP-27752-operator-manages-cilium-identities.md

```diff
@@ -0,0 +1,220 @@
+
+## Label changes
+
+
+### Pod
+
+Pod/Deployment label changes result in pod recreation, which follows the same pod creation path previously described.
+
+
+### Namespace
+
+When namespace labels change, Operator and Agents get the update at the same time. Agents wait until Operator processes the namespace label change update and creates all the new CIDs. Agents keep retrying to match CIDs for all their local endpoints until they are received through the CID watcher.
+
+
+## Simultaneous CID management compatibility
+
+Prior to this feature, all agents managed CIDs and operators did not manage CIDs. When upgrading Cilium to enable this feature, during the transition both agents and operators will manage CIDs. As long as Agents are managing CIDs, the CID duplication issue may continue to occur. Once the upgrade is completed agents will no longer manage CIDs.
+
+Simultaneous CID management is only expected to happen as a transitional state during upgrade from Agents managing CIDs to Operator managing CIDs.
+
+
+## CID deduplication
+
+Operator will not create duplicate CIDs, but it will allow for their existence. In case other sources, like Agents, are making duplicate CIDs, Operator will be compatible with tracking multiple CIDs. This is needed to make Operator compatible with Agents managing CIDs at the same time.
+
+Completely deduplicating CIDs could be done with additional changes in Agent. One option is to have Agents always pick CIDs with the lowest integer value in the name, that matches the label set. This way all Agents would always pick the same CID for the same label set. However, this is only a suggestion, and a proper solution will be covered separately, outside this CFP.
+
+
+## KVStore compatibility
+
+**Important note: kvstore compatibility is a hard requirement for releasing the feature "operator manages CIDs".**
+
+[KVStore](https://docs.cilium.io/en/stable/kvstore/) is a key-value store that is mostly used for sharing Cilium mappings between IPs, identities (IPCache) and labels, across the cluster ([what's inside Cilium’s KVStore](http://arthurchiao.art/blog/whats-inside-cilium-etcd/)).
+
+Operator managing CIDs is intended to work with both `identity-allocation-mode=kvstore` and `identity-allocation-mode=crd`. CRD being the default, the initial implementation will use the CRD allocation mode, and later expanded to other allocation modes.
+
+All features that rely on using kvstore will not work until CID controller is expanded to work with kvstore allocation mode. The work to enable kvstore compatibility is estimated to of medium size and it will be tracked in a separate GitHub issue (https://github.com/cilium/cilium/issues/34865).
+
+
+# Enablement
+
+The feature will be guarded by a flag `operator-manages-identities`, that can be set in container arguments in Operator’s and Agent’s manifests. The default values for both will be `false`, meaning that Operator doesn’t manage CIDs, while Agents do.
```

Comment:
For the record both the agent and operator read the cilium-config config map for configuration. Typically for options shared by the two components we set a single flag value and rely on both components reading the same flag.

If we intend for this to be managed differently for agent vs. operator then we probably want to come up with two separate flags, or alternatively a single flag that represents all possible configurations. For instance:

- `--identity-management-mode=agent`
- `--identity-management-mode=operator`
- `--identity-management-mode=both`

Reply:
Thanks for the suggestion. I added it to the CFP.

Related changes and discussion about enablement is here: https://github.com/cilium/cilium/pull/36411